lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/ReferenceDatasetUtils.scala

```diff
@@ -0,0 +1,144 @@
+      log.info(s"Creating reference training dataset, config:$datasetParams")
+      // Generate the dataset for features
+      val datasetVoidPtr = lightgbmlib.voidpp_handle()
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetCreateFromSampledColumn(
+        sampledData.getSampleData,
+        sampledData.getSampleIndices,
+        numCols,
+        sampledData.getRowCounts,
+        sampledData.numRows,
+        1, // Used for allocation and must be > 0, but we don't use this reference set for data collection
+        numRows,
+        datasetParams,
+        datasetVoidPtr), "Dataset create from samples")
+
+
+      // Serialize the raw dataset to a native buffer
+      val datasetHandle: SWIGTYPE_p_void = lightgbmlib.voidpp_value(datasetVoidPtr)
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetSerializeReferenceToBinary(
+        datasetHandle,
+        bufferHandlePtr,
+        lenPtr), "Serialize ref")
+      val bufferLen: Int = lightgbmlib.intp_value(lenPtr)
+      log.info(s"Created serialized reference dataset of length $bufferLen")
+
+      // The dataset is now serialized to a buffer, so we don't need original
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetFree(datasetHandle), "Free Dataset")
+
+      // This will also free the buffer
+      convertToByteArray(bufferHandlePtr, bufferLen)
+    }
+    finally {
+      sampledData.delete()
+      lightgbmlib.delete_voidpp(datasetVoidPtr)
+      lightgbmlib.delete_voidpp(bufferHandlePtr)
+      lightgbmlib.delete_intp(lenPtr)
+    }
+  }
+
+  def getInitializedReferenceDataset(ctx: PartitionTaskContext): LightGBMDataset = {
+    // The definition is broadcast from Spark, so retrieve it
+    val serializedDataset: Array[Byte] = ctx.trainingCtx.serializedReferenceDataset.get
+
+    // Convert byte array to actual dataset
+    val count = ctx.executorRowCount
+    val datasetParams = ctx.trainingCtx.datasetParams
+    val lightGBMDataset = deserializeReferenceDataset(
+      serializedDataset,
+      count,
+      0,
+      datasetParams)
+
+    // Initialize the dataset for streaming (allocates arrays mostly)
+    val maxOmpThreads = ctx.trainingParams.executionParams.maxStreamingOMPThreads
+    LightGBMUtils.validate(lightgbmlib.LGBM_DatasetInitStreaming(lightGBMDataset.datasetPtr,
+      ctx.trainingCtx.hasWeightsAsInt,
+      ctx.trainingCtx.hasInitialScoresAsInt,
+      ctx.trainingCtx.hasGroupsAsInt,
+      ctx.trainingParams.getNumClass,
+      ctx.executorPartitionCount,
+      maxOmpThreads),
+      "LGBM_DatasetInitStreaming")
+
+    lightGBMDataset.setFeatureNames(ctx.trainingCtx.featureNames, ctx.trainingCtx.numCols)
+    lightGBMDataset
+  }
+
+  private def convertToByteArray(buffer: SWIGTYPE_p_p_void, bufferLen: Int): Array[Byte] = {
+    val byteArray = new Array[Byte](bufferLen)
+    val valPtr = lightgbmlib.new_bytep()
+    val bufferHandle = lightgbmlib.voidpp_value(buffer)
+
+    try
+    {
+      (0 until bufferLen).foreach(i => {
```

Comment:
`foreach { i => ...` (no need for `(i => {...})`)

Reply:
this does not compile. There are 2 lines in there.

Reply:
It should be like this:

```scala
(0 until bufferLen).foreach { i =>
  LightGBMUtils.validate(lightgbmlib.LGBM_ByteBufferGetAt(bufferHandle, i, valPtr), "Buffer getat")
  byteArray(i) = lightgbmlib.bytep_value(valPtr).toByte
}
```

Which part does not compile? Can you show the compilation error?

The reason I insist on changing this part is that using `(i => {...})` is not Scala idiomatic IMHO.

Reply:
value byteArray is not a member of Unit
possible cause: maybe a semicolon is missing before `value byteArray'?
        byteArray(i) = lightgbmlib.bytep_value(valPtr).toByte

Reply:
ah, ok, I figured it out. You moved the bracket to in front of the loop var, which was a subtle change I missed. :)  works now.